libs/utils/src/properties.c

```diff
@@ -103,12 +131,265 @@ static void updateBuffers(char **key, char ** value, char **output, int outputPo
+    } else /*string value*/ {
+        assert(prototype->valueType == CELIX_PROPERTIES_VALUE_TYPE_STRING);
+        entry->value = celix_properties_createString(properties, prototype->typed.strValue);
+        entry->typed.strValue = entry->value;
+    }
+
+    if (entry->value == NULL) {
+        return CELIX_ENOMEM;
+    }
+    return CELIX_SUCCESS;
+}
+
+/**
+ * Allocate entry and optionally use the short properties optimization entries buffer.
+ */
+celix_properties_entry_t* celix_properties_allocEntry(celix_properties_t* properties) {
+    celix_properties_entry_t* entry;
+    if (properties->currentEntriesBufferIndex < CELIX_PROPERTIES_OPTIMIZATION_ENTRIES_BUFFER_SIZE) {
+        entry = &properties->entriesBuffer[properties->currentEntriesBufferIndex++];
+    } else {
+        entry = malloc(sizeof(*entry));
+    }
+    return entry;
+}
+
+/**
+ * Create entry and optionally use the short properties optimization entries buffer and take ownership of the
+ * provided key and value strings.
+ */
+static celix_properties_entry_t* celix_properties_createEntryWithNoCopy(celix_properties_t* properties,
+                                                                        const char* strValue) {
+    celix_properties_entry_t* entry = celix_properties_allocEntry(properties);
+    if (entry == NULL) {
+        return NULL;
+    }
+    entry->value = strValue;
+    entry->valueType = CELIX_PROPERTIES_VALUE_TYPE_STRING;
+    entry->typed.strValue = strValue;
+    return entry;
+}
+
+/**
+ * Create entry and optionally use the short properties optimization buffers.
+ * Only 1 of the types values (strValue, LongValue, etc) should be provided.
+ */
+static celix_properties_entry_t* celix_properties_createEntry(celix_properties_t* properties,
+                                                              const celix_properties_entry_t* prototype) {
+    celix_properties_entry_t* entry = celix_properties_allocEntry(properties);
+    if (entry == NULL) {
+        if (prototype->valueType == CELIX_PROPERTIES_VALUE_TYPE_VERSION) {
+            celix_version_destroy((celix_version_t*)prototype->typed.versionValue);
+        }
+        celix_err_pushf("Cannot allocate property entry");
+        return NULL;
+    }
+
+    celix_status_t status = celix_properties_fillEntry(properties, entry, prototype);
+    if (status != CELIX_SUCCESS) {
+        celix_err_pushf("Cannot fill property entry");
+        if (prototype->valueType == CELIX_PROPERTIES_VALUE_TYPE_VERSION) {
+            celix_version_destroy((celix_version_t*)prototype->typed.versionValue);
+        }
+        if (entry >= properties->entriesBuffer &&
+            entry <= (properties->entriesBuffer + CELIX_PROPERTIES_OPTIMIZATION_ENTRIES_BUFFER_SIZE)) {
+            // entry is part of the properties entries buffer -> decrease the currentEntriesBufferIndex
+            properties->currentEntriesBufferIndex -= 1;
```

Comment:
Yes, this fixed the issue. 
I just noticed we have duplicated code of entry deallocation here and in `celix_properties_destroyEntry`.
To destroy an entry, we must deinitialise then deallocate it. 
In the `celix_properties_deallocEntry`, we can check if the entry is the last one. 
If it is indeed the last one in pool, then we can perform `currentEntriesBufferIndex -= 1`.

This consideration may/may not also apply to https://github.com/apache/celix/pull/470#discussion_r1397781840
So that we don't introduce much complexity.